demux/demux_mkv.c

```diff
@@ -1522,10 +1522,12 @@ static int demux_mkv_open_video(demuxer_t *demuxer, mkv_track_t *track)
 
     sh_v->crop.x0 = track->v_crop_left_set ? track->v_crop_left : 0;
     sh_v->crop.y0 = track->v_crop_top_set ? track->v_crop_top : 0;
-    sh_v->crop.x1 = track->v_width -
+    if (track->v_crop_right_set || track->v_crop_bottom_set) {
```

Comment:
This also doesn't work, because you can end up with crop `0x0+X+Y` which is invalid. Could you explain why you are trying to change the cropping logic here?

Reply:
Because this is what's causing #12680. That file doesn't have crop metadata. It stupidly has both `width` and `Original width`. Because `track->v_width` isn't the same as `track->v_dwidth` (ditto for the height), it causes a crop to be applied. ffplay, vlc and mpv before 338f31e0b694eec2f250b62cf7707ec9bdf71daf use the 1920x1080 dimensions which is what I presume is desired.

Basically I'm just trying to keep everything 0 if there's no actual crop set. I'm assuming that's valid (seems like it should be).

Reply:
> Because this is what's causing https://github.com/mpv-player/mpv/issues/12680.

Well, yes. But the code is correct. Container is lying, so there is no good way to handle that. Like I said we can add an option to ignore container and prefer bit-stream metadata, but we wouldn't make it default anyway, so the user for broken files can force ignore container by `--video-crop=0x0+0+0`

> ffplay, vlc 

Both of them probably cannot even crop mkv. So it is hardly an argument.

> Basically I'm just trying to keep everything 0 if there's no actual crop set. I'm assuming that's valid (seems like it should be).

It is valid, because I made it so, but (0, 0, w, h) is also valid crop and not doing that is just hacky. Imagine that container has both incorrect PixelWidth and crop, you couldn't set to 0,0,0,0

If we really care about such files I would prefer similar option `--video-size-method=<bitstream|container>` instead of hacking it in demuxer. And handle the same way as `--video-aspect-method` in `fix_image_params`

Reply:
I'm not saying the code is wrong, but it assumes that the file doesn't have broken tags for the calculation. If the container in question doesn't have any actual crop (like that issue), why bother trying to match `(0, 0, w, h)`? It can lead to undesirable results. If we don't need to trust the metadata, then why have the extra step?

I guess this is my fault. I didn't realize the internal representation for "no crop" for `(0, 0, w, h)`. I thought it was all zeros. That does make this not the right solution I suppose.

Reply:
> I guess this is my fault. I didn't realize the internal representation for "no crop" in image params was (0, 0, w, h). I thought it was all zeros. That does make this not the right solution I suppose.

All zeros are also fine, mostly for the compatibility with "no-crop" case like you said. Both should be equivalent.

> If we don't need to trust the metadata, then why have the extra step?

I just don't like supporting broken files, because this embraces them and let users generate more of broken files. I'm fine with setting it to (0,0,0,0) I'm just voicing my concern that it is not the best solution and will still break if any of the crop parameters are available.

Reply:
Well we can't stop people from making broken files as much as I'd like to.

>I'm just voicing my concern that it is not the best solution and will still break if any of the crop parameters are available.

If you have a file with totally broken crop parameters and width/heights, I think that one is just going to have to be on you. `--video-crop` still still override it. However while we're on the topic does it make sense to use `dw` and `dh` here instead of `track->v_width/track->v_height` directly? I don't remember my mkv specs but displaywidth is supposed to be preferred right?

Reply:
> However while we're on the topic does it make sense to use dw and dh here instead of track->v_width/track->v_height directly? I don't remember my mkv specs but displaywidth is supposed to be preferred right?

Crop is suppose to be applied to pixel width and display width applied on top. See also https://github.com/mpv-player/mpv/pull/12683 because I just noticed that yesterday.

Like I said I'm fine with setting it to 0,0,0,0 if none crop is set in container. It just produces two different values that should be treated the same. Similarly I would need to rewrite #12683 to take that into account. But if we remember about this throughout the code it should be ok.

Reply:
OK I reworked this part a bit which should do the trick I think.